test/bats/test.bats

```diff
@@ -64,20 +69,32 @@ teardown() {
   assert_success
 }
 
-@test "required labels test" {
+@test "required labels dryrun test" {
   run kubectl apply -f ${BATS_TESTS_DIR}/templates/k8srequiredlabels_template.yaml
   assert_success
 
   wait_for_process $WAIT_TIME $SLEEP_TIME "kubectl -n gatekeeper-system wait --for condition=established --timeout=60s crd/k8srequiredlabels.constraints.gatekeeper.sh"
 
-  run kubectl apply -f ${BATS_TESTS_DIR}/constraints/all_ns_must_have_gatekeeper.yaml
+  run kubectl apply -f ${BATS_TESTS_DIR}/constraints/all_ns_must_have_gatekeeper-dryrun.yaml
   assert_success
 
   wait_for_process $WAIT_TIME $SLEEP_TIME "kubectl get k8srequiredlabels.constraints.gatekeeper.sh ns-must-have-gk -o yaml | grep 'id: gatekeeper-controller-manager'"
 
   run kubectl apply -f ${BATS_TESTS_DIR}/good/good_ns.yaml
   assert_success
 
+  wait_for_process $WAIT_TIME $SLEEP_TIME "kubectl get k8srequiredlabels.constraints.gatekeeper.sh ns-must-have-gk -o json | jq '.spec.enforcementAction' | grep dryrun"
+
+  # deploying a violation with dryrun enforcement action will be accepted
+  run kubectl apply -f ${BATS_TESTS_DIR}/bad/bad_ns.yaml
+  assert_success
+
+  run kubectl delete -f ${BATS_TESTS_DIR}/bad/bad_ns.yaml
+  assert_success
+
+  run kubectl apply -f ${BATS_TESTS_DIR}/constraints/all_ns_must_have_gatekeeper.yaml
```

Comment:
There is a race condition here as Gatekeeper may not have ingested the constraint by the time we get to this rule.

If we invert the order of these tests: load deny rule, wait for ingest, affirm deny, load dryrun rule, wait for deny to no longer happen (with timeout), then we avoid race conditions and also make a stronger case that it is dry run, not a failure enforcing the constraint, that allowed the create request through.

Reply:
reverted order